llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -8285,6 +8285,15 @@ static std::optional<bool> isImpliedCondICmps(const ICmpInst *LHS,
   if (areMatchingOperands(L0, L1, R0, R1, AreSwappedOps))
     return isImpliedCondMatchingOperands(LPred, RPred, AreSwappedOps);
 
+  if (RPred == ICmpInst::ICMP_SLT) {
+    Value *X;
+    Value *Y;
+    // x -nsw y is non-negative when x >= y or negative when x < y
+    if (match(R0, m_NSWSub(m_Value(X), m_Value(Y))) && match(R1, m_AllOnes()) &&
+        areMatchingOperands(L0, L1, X, Y, AreSwappedOps))
+      return isImpliedCondMatchingOperands(LPred, RPred, AreSwappedOps);
```

Comment:
It is incorrect that `X <s Y` implies `X -nsw Y <s -1`.

Reply:
Oh, you are right. 
I missed the big premise, and valuetracking can be used except `abs`, so `X -nsw Y <s -1` and `X -nsw Y <=s -1` is unequivalent.

Reply:
Thanks, now I restrict it `x -nsw y <= -1 <==> x < y`.
And add extra match pattern `x > y and C <= 0 ==> x -nsw y < C is false` to fix #54735